<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.7.32">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Hierarchical Cluster Analysis – PSYC559: Applied Machine Learning in Psychology</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js" type="module"></script>
<script src="site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-dark-2fef5ea3f8957b3e4ecc936fc74692ca.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-121a38a6e2c75e451ef209d5f681333d.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

<link rel="stylesheet" href="styles.css">
</head>

<body class="nav-sidebar docked quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./week6_1.html">Dimension Reduction</a></li><li class="breadcrumb-item"><a href="./week6_2.html">Hierarchical Cluster Analysis</a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation docked overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">PSYC559: Applied Machine Learning in Psychology</a> 
    </div>
      </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./introduction.qmd" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Table of Contents</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">Introduction to R</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./1_1.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Basic R Programming</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./1_1b.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">R Packages</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./1_2.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Reading in Data</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./week1_3.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Describing Data Numerically</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./week1_4.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Describing Data Visually</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true">
 <span class="menu-text">Introduction to Machine Learning</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./week2_1.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Important Terms and Distinctions</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./week2_2.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Problem Classes in ML</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./week2_3.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Data Splitting</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./week2_4.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Bias and Variance</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true">
 <span class="menu-text">Feature Engineering</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./week2_5a.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Overview</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./week2_5b.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Example Data</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./week2_5c.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Missing Data</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./week2_5d.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Feature Filtering</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./week2_5e.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Numeric Feature Engineering</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./week2_5f.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Categorical Feature Engineering</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./week2_5g.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Workflow</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="true">
 <span class="menu-text">Regression Methods</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./week3_1.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Simple Linear Regression</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./week3_2.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Multiple Linear Regression</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./week3_3.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Simple Logistic Regression</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./week3_4.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Multiple Logistic Regression</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./week4_1.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Regularized Regression</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" role="navigation" aria-expanded="true">
 <span class="menu-text">Tree Methods</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./week5_1.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Decision Trees</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./week5_2.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Bagging</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./week5_3.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Random Forests</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./week5_4.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Gradient Boosting</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" role="navigation" aria-expanded="true">
 <span class="menu-text">Dimension Reduction</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-6" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./week6_1.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Principal Components Analysis</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./week6_2.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">Hierarchical Cluster Analysis</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./week6_3.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">K-Means Cluster Analysis</span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#introduction" id="toc-introduction" class="nav-link active" data-scroll-target="#introduction">Introduction</a></li>
  <li><a href="#supervised-learning" id="toc-supervised-learning" class="nav-link" data-scroll-target="#supervised-learning">Supervised Learning</a></li>
  <li><a href="#unsupervised-learning" id="toc-unsupervised-learning" class="nav-link" data-scroll-target="#unsupervised-learning">Unsupervised Learning</a></li>
  <li><a href="#cluster-analysis" id="toc-cluster-analysis" class="nav-link" data-scroll-target="#cluster-analysis">Cluster Analysis</a></li>
  <li><a href="#clustering-algorithms" id="toc-clustering-algorithms" class="nav-link" data-scroll-target="#clustering-algorithms">Clustering Algorithms</a></li>
  <li><a href="#hierarchical-clustering" id="toc-hierarchical-clustering" class="nav-link" data-scroll-target="#hierarchical-clustering">Hierarchical Clustering</a></li>
  <li><a href="#k-means-clustering" id="toc-k-means-clustering" class="nav-link" data-scroll-target="#k-means-clustering">K-means Clustering</a></li>
  <li><a href="#hierarchical-clustering-1" id="toc-hierarchical-clustering-1" class="nav-link" data-scroll-target="#hierarchical-clustering-1">Hierarchical Clustering</a></li>
  <li><a href="#distances" id="toc-distances" class="nav-link" data-scroll-target="#distances">Distances</a></li>
  <li><a href="#distance-between-clusters" id="toc-distance-between-clusters" class="nav-link" data-scroll-target="#distance-between-clusters">Distance Between Clusters</a></li>
  <li><a href="#linkages-applied-example" id="toc-linkages-applied-example" class="nav-link" data-scroll-target="#linkages-applied-example">Linkages Applied Example</a></li>
  <li><a href="#dendograms" id="toc-dendograms" class="nav-link" data-scroll-target="#dendograms">Dendograms</a></li>
  <li><a href="#interpreting-dendograms" id="toc-interpreting-dendograms" class="nav-link" data-scroll-target="#interpreting-dendograms">Interpreting Dendograms</a></li>
  <li><a href="#identifying-clusters-in-dendograms" id="toc-identifying-clusters-in-dendograms" class="nav-link" data-scroll-target="#identifying-clusters-in-dendograms">Identifying Clusters in Dendograms</a></li>
  <li><a href="#hierarchical-clustering-in-r" id="toc-hierarchical-clustering-in-r" class="nav-link" data-scroll-target="#hierarchical-clustering-in-r">Hierarchical Clustering in R</a></li>
  <li><a href="#example-data" id="toc-example-data" class="nav-link" data-scroll-target="#example-data">Example Data</a></li>
  <li><a href="#preparing-data" id="toc-preparing-data" class="nav-link" data-scroll-target="#preparing-data">Preparing Data</a></li>
  <li><a href="#scaling" id="toc-scaling" class="nav-link" data-scroll-target="#scaling">Scaling</a></li>
  <li><a href="#plotting" id="toc-plotting" class="nav-link" data-scroll-target="#plotting">Plotting</a></li>
  <li><a href="#clustering-algorithms-1" id="toc-clustering-algorithms-1" class="nav-link" data-scroll-target="#clustering-algorithms-1">Clustering Algorithms</a></li>
  <li><a href="#agglomerative-clustering" id="toc-agglomerative-clustering" class="nav-link" data-scroll-target="#agglomerative-clustering">Agglomerative Clustering</a></li>
  <li><a href="#divisive-hierarchical-clustering" id="toc-divisive-hierarchical-clustering" class="nav-link" data-scroll-target="#divisive-hierarchical-clustering">Divisive hierarchical clustering</a></li>
  <li><a href="#determining-the-number-of-clusters" id="toc-determining-the-number-of-clusters" class="nav-link" data-scroll-target="#determining-the-number-of-clusters">Determining the number of clusters</a></li>
  <li><a href="#dendograms-1" id="toc-dendograms-1" class="nav-link" data-scroll-target="#dendograms-1">Dendograms</a></li>
  <li><a href="#final-thoughts" id="toc-final-thoughts" class="nav-link" data-scroll-target="#final-thoughts">Final Thoughts</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./week6_1.html">Dimension Reduction</a></li><li class="breadcrumb-item"><a href="./week6_2.html">Hierarchical Cluster Analysis</a></li></ol></nav>
<div class="quarto-title">
<h1 class="title">Hierarchical Cluster Analysis</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<section id="introduction" class="level1" style="color: yellow">
<h1>Introduction</h1>
</section>
<p>This section of the course book covers the basics of cluster analysis. First, we introduce the foundational topics and then proceed with applied examples in R.</p>
<section id="supervised-learning" class="level2" style="color: orange">
<h2 class="anchored" data-anchor-id="supervised-learning">Supervised Learning</h2>
</section>
<p><strong>Objective: Prediction</strong></p>
<ul>
<li>Access to a set of variables, <span class="math inline">\(x_1,x_2,\dots,x_p\)</span>, measured on <span class="math inline">\(n\)</span> observations, and a response <span class="math inline">\(y\)</span> also measured on those same <span class="math inline">\(n\)</span> observations.</li>
<li>Objective is to predict <span class="math inline">\(y\)</span> using <span class="math inline">\(x_1,x_2,\dots,x_p\)</span></li>
</ul>
<p>In many common situations there are a well-developed set of tools for supervised learning:</p>
<ul>
<li>Regression and classification
<ul>
<li>logistic regression, trees, random forests, bagging, boosting</li>
</ul></li>
<li>Clear understanding of how to assess the quality of obtained results
<ul>
<li>cross-validation, model fit, etc.</li>
</ul></li>
</ul>
<section id="unsupervised-learning" class="level2" style="color: orange">
<h2 class="anchored" data-anchor-id="unsupervised-learning">Unsupervised Learning</h2>
</section>
<p><strong>Objective: Description</strong></p>
<ul>
<li>A set of statistical tools intended for when we have a set of variables, <span class="math inline">\(x_1,x_2,\dots,x_p\)</span>, measured on <span class="math inline">\(n\)</span> observations.</li>
<li>Objective is to uncover interesting patterns in the measurements of <span class="math inline">\(x_1,x_2,\dots,x_p\)</span>
<ul>
<li>Not interested in prediction because we do not have an associated response variable <span class="math inline">\(y\)</span></li>
<li>Can we discover subgroups among the variables or among the observations?</li>
<li>Is there an informative way to visualize the data?</li>
</ul></li>
</ul>
<p>Unsupervised learning is, in some ways, more challenging than supervised learning.</p>
<ul>
<li>Tends to be more subjective.</li>
<li>No simple goal for the analysis, such as prediction of a response</li>
<li>Often hard to assess the results obtained from unsupervised learning methods because there is no universally accepted mechanism for validating the results</li>
<li>No way to check our work because we do not know the true answer—the problem is unsupervised</li>
</ul>
<p>The importance of reliable unsupervised learning methods is growing in a number of fields:</p>
<ul>
<li>A cancer researcher might assay gene expression levels in patients with breast cancer to look for subgroups among the breast cancer samples in order to obtain a better understanding of the disease</li>
<li>An online shopping site might try to identify groups of shoppers with similar browsing and purchase histories to target coupons, sales, etc.</li>
<li>A search engine might choose what search results to display to a particular individual based on the click histories of other individuals with similar search patterns.</li>
</ul>
<section id="cluster-analysis" class="level1" style="color: yellow">
<h1>Cluster Analysis</h1>
</section>
<p>Cluster analysis or clustering is the task of grouping a set of objects in such a way that objects in the same group (a cluster) are more similar (in some sense or another) to each other than to those in other groups (clusters).</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="imgs/cluster.png" class="img-fluid figure-img"></p>
<figcaption><a href="https://en.wikipedia.org/wiki/Cluster_analy" class="uri">https://en.wikipedia.org/wiki/Cluster_analy</a></figcaption>
</figure>
</div>
<p>We seek to partition observations into distinct groups so that the observations within each group are quite similar to each other, while observations in different groups are quite different from each other.</p>
<p>To make this concrete, we must define what it means for two or more observations to be similar or different</p>
<ul>
<li>most often this is done by the measurement of distance</li>
<li>cluster analysis methods work from <em>dissimilarity</em> measures (e.g., distance matrix)</li>
</ul>
<section id="person-oriented-clustering" class="level4" style="color: lightyellow">
<h4 class="anchored" data-anchor-id="person-oriented-clustering">Person-Oriented Clustering</h4>
</section>
<p>We can think of cluster analysis as a person-oriented approach in that one objective of cluster analysis is to identify different <em>types of people</em>.</p>
<section id="clustering-algorithms" class="level1" style="color: yellow">
<h1>Clustering Algorithms</h1>
</section>
<p>Many clustering algorithms exist:</p>
<ol type="1">
<li>Ward’s hierarchical clustering</li>
<li>K-means clustering</li>
<li>Hierarchical clustering</li>
<li>Density-based clustering</li>
<li>Spectral clustering</li>
<li>Mean shift</li>
<li>Affinity propagation</li>
<li>Mixture model (latent profile analysis, latent class analysis)</li>
</ol>
<p>The majority of class today will cover (1) Hierarchical Clustering.</p>
<section id="hierarchical-clustering" class="level2" style="color: orange">
<h2 class="anchored" data-anchor-id="hierarchical-clustering">Hierarchical Clustering</h2>
</section>
<ul>
<li>We do not know in advance how many clusters we want</li>
<li>Bottom-up approach (grouping similar observations together)</li>
<li>End up with a tree-like visual representation of the observations called a dendrogram</li>
</ul>
<section id="k-means-clustering" class="level2" style="color: orange">
<h2 class="anchored" data-anchor-id="k-means-clustering">K-means Clustering</h2>
</section>
<p>K-means clustering:</p>
<ul>
<li>Seek to partition the observations into a pre-specified number of clusters</li>
<li>Top-down approach</li>
</ul>
<section id="hierarchical-clustering-1" class="level1" style="color: yellow">
<h1>Hierarchical Clustering</h1>
</section>
<p>Hierarchical clustering proceeds via an extremely simple algorithm:</p>
<p>Begin by defining some sort of dissimilarity measure between each pair of observations (e.g., Euclidean distance).</p>
<p>Algorithm proceeds iteratively:</p>
<ol type="1">
<li>Each of the <span class="math inline">\(n\)</span> observations is treated as its own cluster</li>
<li>The two clusters that are most similar are then fused so that there now are <span class="math inline">\(n-1\)</span> clusters</li>
<li>Repeat Step 2 until all of the observations belong to one single cluster</li>
</ol>
<p>Note this produces not one clustering, but a family of clustering represented by a dendrogram.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="imgs/hclust.jpg" class="img-fluid figure-img"></p>
<figcaption>Hierarchical Clustering</figcaption>
</figure>
</div>
<section id="distances" class="level2" style="color: orange">
<h2 class="anchored" data-anchor-id="distances">Distances</h2>
</section>
<p>Euclidean distance is a common distance measure. For two dimensions, it is equal to the sum of squares of difference on <span class="math inline">\(x\)</span> plus the sum of squares of difference on <span class="math inline">\(y\)</span>. Note, variables can differ in scale so it is important to standardize our inputs.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="imgs/euclidian.jpg" class="img-fluid figure-img"></p>
<figcaption>Euclidian Distance</figcaption>
</figure>
</div>
<p>There are many <strong>distance measures</strong>.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="imgs/manhattan.jpg" class="img-fluid figure-img"></p>
<figcaption>Distance Measures</figcaption>
</figure>
</div>
<p>A taxicab geometry is a form of geometry in which the usual distance function or metric of Euclidean geometry is replaced by a new metric in which the distance between two points is the sum of the absolute differences of their Cartesian coordinates.</p>
<p>Taxicab geometry versus Euclidean distance: In taxicab geometry, the red, yellow, and blue paths all have the same shortest path length of 12.</p>
<section id="distance-between-clusters" class="level3" style="color: lightyellow">
<h3 class="anchored" data-anchor-id="distance-between-clusters">Distance Between Clusters</h3>
</section>
<p>The concept of dissimilarity between a pair of observations needs to be extended to a pair of groups of observations – what’s the distance between clusters?</p>
<p>This extension is achieved by developing the notion of <strong>linkage</strong>, which defines the dissimilarity between two groups of observations. Four most common types of linkage: complete, average, single, and centroid. An important fifth is the Ward-method.</p>
<p><strong>Complete</strong>: Maximal intercluster dissimilarity. Compute all pairwise dissimilarities between the observations in cluster A and the observations in cluster B, and record the largest of these dissimilarities. This is sometimes referred to as farthest neighbor clustering.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="imgs/linkage_complete.jpg" class="img-fluid figure-img"></p>
<figcaption>Complete Linkage</figcaption>
</figure>
</div>
<p><strong>Single</strong>: Minimal intercluster dissimilarity. Compute all pairwise dissimilarities between the observations in cluster A and the observations in cluster B, and record the smallest of these dissimilarities.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="imgs/linkage_single.jpg" class="img-fluid figure-img"></p>
<figcaption>Single Linkage</figcaption>
</figure>
</div>
<p><strong>Average</strong>: Mean intercluster dissimilarity. Compute all pairwise dissimilarities between the observations in cluster A and the observations in cluster B, and record the average of these dissimilarities.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="imgs/linkage_average.jpg" class="img-fluid figure-img"></p>
<figcaption>Average Linkage</figcaption>
</figure>
</div>
<p><strong>Centroid</strong>: Dissimilarity between the centroid for cluster A (a mean vector of length p) and the centroid for cluster B.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="imgs/linkage_centroid.jpg" class="img-fluid figure-img"></p>
<figcaption>Centroid Linkage</figcaption>
</figure>
</div>
<p><strong>Ward’s Method</strong>: minimize within cluster sum of squares. The linkage function specifying the distance between two clusters is computed as the increase in the error sum of squares after fusing two clusters into a single cluster.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="imgs/linkage_ward.jpg" class="img-fluid figure-img"></p>
<figcaption>Centroid Linkage</figcaption>
</figure>
</div>
<section id="linkages-applied-example" class="level3" style="color: lightyellow">
<h3 class="anchored" data-anchor-id="linkages-applied-example">Linkages Applied Example</h3>
</section>
<p>Let’s take a look at different linkages using this US cities example. First we can calculate the difference among the different cities</p>
<p><img src="imgs/clust_slide1.jpg" class="img-fluid" alt="US Map"> <img src="imgs/clust_slide2.jpg" class="img-fluid" alt="Matrix of Dissimilarity"></p>
<p>Now, let’s start by making a cluster of Boston and New York.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="imgs/clust_slide3.jpg" class="img-fluid figure-img"></p>
<figcaption>US Map</figcaption>
</figure>
</div>
<p>Now compute new dissimilarity matrix. New matrix depends on the linkage approach.</p>
<ul>
<li>Complete linkage would use distance from Boston to all other cities because Boston is further than NY</li>
<li>Single linkage would use distance from NY to all other cities because NY is closer than Boston</li>
<li>Average linkage would use the mean of the distances from NY and Boston</li>
<li>Centroid linkage would place a new city half way between NY and Boston and calculate differences from this new city</li>
</ul>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="imgs/clust_slide4.jpg" class="img-fluid figure-img"></p>
<figcaption>Resulting Clusters</figcaption>
</figure>
</div>
<section id="dendograms" class="level2" style="color: orange">
<h2 class="anchored" data-anchor-id="dendograms">Dendograms</h2>
</section>
<p>Now let’s look at the dendograms.</p>
<p><img src="imgs/dend_single.png" class="img-fluid" alt="Dendogram: Single Linkage"> <img src="imgs/dend_complete.png" class="img-fluid" alt="Dendogram: Complete Linkage"> Which one gives us a better solution for the underlying typology?</p>
<section id="interpreting-dendograms" class="level3" style="color: lightorange">
<h3 class="anchored" data-anchor-id="interpreting-dendograms">Interpreting Dendograms</h3>
</section>
<ul>
<li>Each leaf of the dendrogram represents one of the <span class="math inline">\(n\)</span> observations</li>
<li>Moving up the tree, leaves begin to fuse into branches corresponding to observations that are similar to each other</li>
<li>Moving higher up, branches fuse, either with leaves or other branches</li>
<li>The height of the fusion, as measured on the vertical axis, indicates how different the two observations are</li>
<li>Observations that fuse at the very bottom of the tree are quite similar to each other</li>
<li>Observations that fuse close to the top of the tree will tend to be quite different</li>
</ul>
<p>We can draw conclusions about the similarity of two observations based on the location on the vertical axis where branches containing those two observations first are fused.</p>
<p>Generally we are looking for a level above which the lines are long (between-group heterogeneity) and below which the leaves are close (within-group homogeneity).</p>
<section id="identifying-clusters-in-dendograms" class="level3" style="color: lightorange">
<h3 class="anchored" data-anchor-id="identifying-clusters-in-dendograms">Identifying Clusters in Dendograms</h3>
</section>
<ul>
<li>Make a horizontal cut across the dendrogram</li>
<li>The distinct sets of observations beneath the cut can be interpreted as clusters</li>
<li>Therefore, a dendrogram can be used to obtain any number of clusters</li>
</ul>
<p>Researchers often look at the dendrogram and select by eye a sensible number of clusters based on the heights of the fusion and the number of clusters desired.</p>
<p>The choice of where to cut the dendrogram is not always clear.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="imgs/dend3.jpg" class="img-fluid figure-img"></p>
<figcaption>Example Dendogram</figcaption>
</figure>
</div>
<p>What should we do with this data? 1 2 or 3 groups? You as a researcher have to decide.</p>
<div style="color: lightorange">
<p><strong>Discussion Questions</strong></p>
</div>
<ol type="1">
<li>Identify a dimension or process central to your own interests or research where you think there is a substantial between-person heterogeneity.</li>
<li>Identify a set of variables that might be used to classify individuals into meaningful groups as a means to address this heterogeneity.</li>
<li>With groups or clusters identified how might you empirically assess these groupings?</li>
</ol>
<section id="hierarchical-clustering-in-r" class="level1" style="color: yellow">
<h1>Hierarchical Clustering in R</h1>
</section>
<section id="example-data" class="level2" style="color: orange">
<h2 class="anchored" data-anchor-id="example-data">Example Data</h2>
</section>
<p>Our example makes use of an experience sampling data set, but treats these data as though they are cross-sectional. Getting the data and doing a bit of data management (new id variable)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co">#set filepath for data file</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>filepath <span class="ot">&lt;-</span> <span class="st">"https://quantdev.ssri.psu.edu/sites/qdev/files/AMIBbrief_raw_daily1.csv"</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="co">#read in the .csv file using the url() function</span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>daily <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="at">file=</span><span class="fu">url</span>(filepath),<span class="at">header=</span><span class="cn">TRUE</span>)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="co">#clean-up of variable names so that they are all lowercase</span></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>var.names.daily <span class="ot">&lt;-</span> <span class="fu">tolower</span>(<span class="fu">colnames</span>(daily))</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(daily)<span class="ot">&lt;-</span>var.names.daily</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="co">#creating a new "id" variable</span></span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a><span class="co">#(we had repeated measures nested in people, now they all get different ids)</span></span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a>daily<span class="sc">$</span>id <span class="ot">&lt;-</span> daily<span class="sc">$</span>id<span class="sc">*</span><span class="dv">10</span><span class="sc">+</span>daily<span class="sc">$</span>day</span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(daily)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "id"      "day"     "date"    "slphrs"  "weath"   "lteq"    "pss"    
 [8] "se"      "swls"    "evalday" "posaff"  "negaff"  "temp"    "hum"    
[15] "wind"    "bar"     "prec"   </code></pre>
</div>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co">#reducing down to variable set</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>daily <span class="ot">&lt;-</span> daily[ ,<span class="fu">c</span>(<span class="st">"id"</span>,<span class="st">"slphrs"</span>,<span class="st">"weath"</span>,<span class="st">"lteq"</span>,<span class="st">"pss"</span>,<span class="st">"se"</span>,<span class="st">"swls"</span>,<span class="st">"evalday"</span>, <span class="st">"posaff"</span>,<span class="st">"negaff"</span>,<span class="st">"temp"</span>,<span class="st">"hum"</span>,<span class="st">"wind"</span>,<span class="st">"bar"</span>,<span class="st">"prec"</span>)]</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="co">#names of variables</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(daily)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "id"      "slphrs"  "weath"   "lteq"    "pss"     "se"      "swls"   
 [8] "evalday" "posaff"  "negaff"  "temp"    "hum"     "wind"    "bar"    
[15] "prec"   </code></pre>
</div>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co">#looking at data    </span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(daily,<span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     id slphrs weath lteq  pss se swls evalday posaff negaff temp  hum wind
1  1010    6.0     1   10 2.50  2  3.8       1    3.9    3.0 28.0 0.79 11.0
2  1011    2.0     2   10 2.75  3  4.2       0    3.8    2.3 20.8 0.62  3.6
3  1012    9.0     3   10 3.50  4  5.0       1    5.1    1.0 29.1 0.51  1.9
4  1013    7.5     2    9 3.00  4  5.0       1    5.6    1.3 30.2 0.58  2.7
5  1014    8.0     1   18 2.75  3  4.0       1    4.3    1.1 22.7 0.55  2.4
6  1015    8.0     2   19 2.75  3  4.2       1    3.9    1.0 21.4 0.54  0.7
7  1016    8.0     3   21 3.50  4  4.6       1    5.1    1.2 31.4 0.49  1.0
8  1017    7.0    NA   14 2.75  3  4.6       1    4.8    1.1 45.3 0.52  1.1
9  1020    7.0     0   12 3.50  5  5.6       0    6.3    1.4 28.0 0.79 11.0
10 1021    6.0     0   20 4.00  5  6.6       0    7.0    1.6 20.8 0.62  3.6
     bar prec
1  29.40 0.20
2  30.17 0.00
3  30.35 0.02
4  30.23 0.00
5  30.46 0.00
6  30.54 0.00
7  30.51 0.00
8  30.30 0.00
9  29.40 0.20
10 30.17 0.00</code></pre>
</div>
</div>
<section id="preparing-data" class="level3" style="color: lightyellow">
<h3 class="anchored" data-anchor-id="preparing-data">Preparing Data</h3>
</section>
<p>Note that cluster analysis does NOT generally work with missing data. Here we simply delete incomplete cases. Other possibilities include imputation, and calculation of distances using most complete subsets.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co">#removing observations with NA</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>dailysub <span class="ot">&lt;-</span> daily[<span class="fu">complete.cases</span>(daily), ] </span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a><span class="fu">describe</span>(dailysub)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        vars    n    mean      sd  median trimmed     mad     min     max
id         1 1376 3276.28 1279.88 3271.50 3302.24 1497.43 1010.00 5327.00
slphrs     2 1376    7.20    1.81    7.00    7.20    1.48    0.00   18.00
weath      3 1376    2.00    1.29    2.00    2.00    1.48    0.00    4.00
lteq       4 1376   12.50   10.42    9.00   11.24    8.90    0.00   58.00
pss        5 1376    2.62    0.68    2.75    2.64    0.74    0.00    4.00
se         6 1376    3.43    0.99    3.00    3.46    1.48    1.00    5.00
swls       7 1376    4.11    1.27    4.20    4.15    1.19    1.00    7.00
evalday    8 1376    0.68    0.46    1.00    0.73    0.00    0.00    1.00
posaff     9 1376    4.11    1.10    4.20    4.14    1.19    1.00    7.00
negaff    10 1376    2.45    1.04    2.20    2.34    1.04    1.00    6.90
temp      11 1376   40.18    7.88   42.00   40.51    8.90   20.80   56.00
hum       12 1376    0.62    0.20    0.66    0.63    0.21    0.24    0.90
wind      13 1376    7.36    4.45    7.00    6.81    4.45    0.70   20.00
bar       14 1376   30.02    0.33   30.00   30.04    0.43   29.32   30.54
prec      15 1376    0.05    0.09    0.00    0.03    0.00    0.00    0.30
          range  skew kurtosis    se
id      4317.00 -0.10    -1.04 34.50
slphrs    18.00  0.12     1.93  0.05
weath      4.00 -0.06    -1.06  0.03
lteq      58.00  1.07     0.95  0.28
pss        4.00 -0.37     0.17  0.02
se         4.00 -0.40    -0.12  0.03
swls       6.00 -0.28    -0.22  0.03
evalday    1.00 -0.79    -1.37  0.01
posaff     6.00 -0.24    -0.37  0.03
negaff     5.90  0.95     0.67  0.03
temp      35.20 -0.37    -0.24  0.21
hum        0.66 -0.36    -1.10  0.01
wind      19.30  0.97     0.86  0.12
bar        1.22 -0.39    -0.90  0.01
prec       0.30  1.85     1.98  0.00</code></pre>
</div>
</div>
<section id="scaling" class="level3" style="color: lightyellow">
<h3 class="anchored" data-anchor-id="scaling">Scaling</h3>
</section>
<p>The unit of distance may be different for different variables. For example, one year of difference in age seems like it should be a larger difference than one dollar difference in income.</p>
<p>Different variables will be “weighted” differently in the distance calculation. To alleviate this, a common approach is to rescale each variable into a standardized, <em>z-score</em> variable (i.e., by subtracting the mean and dividing by the standard deviation).</p>
<p>Thus, all the variables would then have mean = 0, with differences scales in standard deviation units. Note that this scales everything in relation to the observed sample (which has plusses and minuses).</p>
<p>The R function <code>scale()</code> makes it all very easy.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co">#scaling all the variables</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>dailyscale <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="fu">scale</span>(dailysub, <span class="at">center=</span><span class="cn">TRUE</span>, <span class="at">scale=</span><span class="cn">TRUE</span>))</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a><span class="co">#checking and fixing the id variable (which we did not want standardized)</span></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(dailyscale<span class="sc">$</span>id)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> num [1:1376] -1.77 -1.77 -1.77 -1.77 -1.77 ...</code></pre>
</div>
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>dailyscale<span class="sc">$</span>id <span class="ot">&lt;-</span> dailysub<span class="sc">$</span>id</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(dailyscale<span class="sc">$</span>id)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> num [1:1376] 1010 1011 1012 1013 1014 ...</code></pre>
</div>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">describe</span>(dailyscale)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        vars    n    mean      sd  median trimmed     mad     min     max
id         1 1376 3276.28 1279.88 3271.50 3302.24 1497.43 1010.00 5327.00
slphrs     2 1376    0.00    1.00   -0.11    0.00    0.82   -3.98    5.98
weath      3 1376    0.00    1.00    0.00    0.00    1.15   -1.55    1.54
lteq       4 1376    0.00    1.00   -0.34   -0.12    0.85   -1.20    4.37
pss        5 1376    0.00    1.00    0.19    0.03    1.08   -3.83    2.02
se         6 1376    0.00    1.00   -0.43    0.04    1.49   -2.45    1.58
swls       7 1376    0.00    1.00    0.07    0.03    0.93   -2.45    2.27
evalday    8 1376    0.00    1.00    0.68    0.10    0.00   -1.47    0.68
posaff     9 1376    0.00    1.00    0.08    0.03    1.08   -2.82    2.63
negaff    10 1376    0.00    1.00   -0.24   -0.11    1.00   -1.40    4.28
temp      11 1376    0.00    1.00    0.23    0.04    1.13   -2.46    2.01
hum       12 1376    0.00    1.00    0.20    0.05    1.06   -1.94    1.42
wind      13 1376    0.00    1.00   -0.08   -0.12    1.00   -1.50    2.84
bar       14 1376    0.00    1.00   -0.05    0.06    1.29   -2.09    1.56
prec      15 1376    0.00    1.00   -0.53   -0.26    0.00   -0.53    2.68
          range  skew kurtosis    se
id      4317.00 -0.10    -1.04 34.50
slphrs     9.97  0.12     1.93  0.03
weath      3.09 -0.06    -1.06  0.03
lteq       5.57  1.07     0.95  0.03
pss        5.85 -0.37     0.17  0.03
se         4.03 -0.40    -0.12  0.03
swls       4.72 -0.28    -0.22  0.03
evalday    2.15 -0.79    -1.37  0.03
posaff     5.45 -0.24    -0.37  0.03
negaff     5.68  0.95     0.67  0.03
temp       4.47 -0.37    -0.24  0.03
hum        3.36 -0.36    -1.10  0.03
wind       4.33  0.97     0.86  0.03
bar        3.65 -0.39    -0.90  0.03
prec       3.21  1.85     1.98  0.03</code></pre>
</div>
</div>
<section id="plotting" class="level3" style="color: lightyellow">
<h3 class="anchored" data-anchor-id="plotting">Plotting</h3>
</section>
<p>We choose a small subset of variables for easy visualization in a bivariate space. We use <code>lteq</code>, a measure of physical activity (Leisure Time Exercise Questionnaire), and <code>posaff</code>, a measure of positive affect.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(dailyscale,<span class="fu">aes</span>(<span class="at">x=</span>lteq,<span class="at">y=</span>posaff)) <span class="sc">+</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="week6_2_files/figure-html/unnamed-chunk-5-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>This is a good toy data set for class purposes, but keep in mind the original nature of the data which might not be the best for cluster analytic purposes.</p>
<section id="clustering-algorithms-1" class="level2" style="color: orange">
<h2 class="anchored" data-anchor-id="clustering-algorithms-1">Clustering Algorithms</h2>
</section>
<p>We will consider two hierarchical clustering algorithms:</p>
<p>1.<strong>Agglomerative clustering</strong>:</p>
<ul>
<li>Commonly referred to as AGNES (AGglomerative NESting)</li>
<li>Bottom- up approachs works in a bottom-up manner.</li>
<li>This is the same procedure we followed above when clustering US cities</li>
<li>Vomplete linkage and Ward’s method are often preferred for AGNES</li>
</ul>
<ol start="2" type="1">
<li><em>Divisive hierarchical clustering</em>:</li>
</ol>
<ul>
<li>Commonly referred to as DIANA (DIvisive ANAlysis)</li>
<li>Top-down approach works in a top-down manner.</li>
<li>For DIANA, clusters are divided based on the maximum average dissimilarity</li>
</ul>
<section id="agglomerative-clustering" class="level3" style="color: lightyellow">
<h3 class="anchored" data-anchor-id="agglomerative-clustering">Agglomerative Clustering</h3>
</section>
<p>To perform agglomerative HC with <code>hclust()</code>, we first compute the dissimilarity values with <code>dist()</code> and then feed these values into <code>hclust()</code> and specify the agglomeration method to be used (i.e.&nbsp;<code>"complete"</code>, <code>"average"</code>, <code>"single"</code>, or <code>"ward.D"</code>).</p>
<p>First, however, let’s remember that when we think about distances we are imagining each individual as a point in a multivariate space. For example, let’s look at the first three individuals.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>data1 <span class="ot">&lt;-</span> dailyscale[<span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">3</span>,<span class="dv">12</span>),<span class="fu">c</span>(<span class="st">"id"</span>,<span class="st">"lteq"</span>,<span class="st">"posaff"</span>)]</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(data1,<span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     id       lteq     posaff
1  1010 -0.2399597 -0.1889168
3  1012 -0.2399597  0.9016850
14 1025 -0.4318716  0.5381510</code></pre>
</div>
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>labels.abc <span class="ot">&lt;-</span><span class="fu">c</span>(<span class="st">"A"</span>,<span class="st">"B"</span>,<span class="st">"C"</span>)</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(data1,<span class="fu">aes</span>(<span class="at">x=</span>lteq,<span class="at">y=</span>posaff)) <span class="sc">+</span></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_polygon</span>(<span class="at">fill=</span><span class="st">"blue"</span>,<span class="at">alpha=</span>.<span class="dv">6</span>) <span class="sc">+</span></span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size=</span><span class="dv">3</span>) <span class="sc">+</span></span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text</span>(<span class="fu">aes</span>(<span class="at">x=</span>lteq<span class="fl">-.1</span>,<span class="at">label=</span>labels.abc)) <span class="sc">+</span></span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylim</span>(<span class="sc">-</span><span class="dv">1</span>,<span class="dv">1</span>) <span class="sc">+</span> <span class="fu">xlim</span>(<span class="sc">-</span><span class="dv">1</span>,<span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="week6_2_files/figure-html/unnamed-chunk-6-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>In <code>hclust</code> we first need to create a dissimilarity matrix. Here we use Euclidean distance and then perform hierarchical clustering using complete linkage.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(cluster)</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a><span class="co"># For reproducibility</span></span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Dissimilarity matrix</span></span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a>d <span class="ot">&lt;-</span> <span class="fu">dist</span>(dailyscale[,<span class="fu">c</span>(<span class="st">"lteq"</span>,<span class="st">"posaff"</span>)], <span class="at">method =</span> <span class="st">"euclidean"</span>)</span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Hierarchical clustering using Complete Linkage</span></span>
<span id="cb19-10"><a href="#cb19-10" aria-hidden="true" tabindex="-1"></a>hc1 <span class="ot">&lt;-</span> <span class="fu">hclust</span>(d, <span class="at">method =</span> <span class="st">"complete"</span> )</span>
<span id="cb19-11"><a href="#cb19-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-12"><a href="#cb19-12" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(hc1, <span class="at">cex =</span> <span class="fl">0.6</span>, <span class="at">hang =</span> <span class="sc">-</span><span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="week6_2_files/figure-html/unnamed-chunk-7-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Alternatively, we can use the <code>agnes()</code> function. This function behaves similar to <code>hclust()</code>; however, with the <code>agnes()</code> function you can also get the agglomerative coefficient (AC), which measures the amount of clustering structure found.</p>
<p>Generally speaking, the closer AC is to 1, the more clustered the data are. An AC above 0.75 is considered strong clustering structure.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="co"># For reproducibility</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute maximum or complete linkage clustering with agnes</span></span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>hc2 <span class="ot">&lt;-</span> <span class="fu">agnes</span>(dailyscale[,<span class="fu">c</span>(<span class="st">"lteq"</span>,<span class="st">"posaff"</span>)], <span class="at">method =</span> <span class="st">"complete"</span>)</span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Agglomerative coefficient</span></span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a>hc2<span class="sc">$</span>ac</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.9938078</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="co"># methods to assess</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>m <span class="ot">&lt;-</span> <span class="fu">c</span>( <span class="st">"average"</span>, <span class="st">"single"</span>, <span class="st">"complete"</span>, <span class="st">"ward"</span>)</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(m) <span class="ot">&lt;-</span> <span class="fu">c</span>( <span class="st">"average"</span>, <span class="st">"single"</span>, <span class="st">"complete"</span>, <span class="st">"ward"</span>)</span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a><span class="co"># function to compute coefficient</span></span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a>ac <span class="ot">&lt;-</span> <span class="cf">function</span>(x) {</span>
<span id="cb22-7"><a href="#cb22-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">agnes</span>(dailyscale[,<span class="fu">c</span>(<span class="st">"lteq"</span>,<span class="st">"posaff"</span>)], <span class="at">method =</span> x)<span class="sc">$</span>ac</span>
<span id="cb22-8"><a href="#cb22-8" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb22-9"><a href="#cb22-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-10"><a href="#cb22-10" aria-hidden="true" tabindex="-1"></a><span class="co"># get agglomerative coefficient for each linkage method</span></span>
<span id="cb22-11"><a href="#cb22-11" aria-hidden="true" tabindex="-1"></a>purrr<span class="sc">::</span><span class="fu">map_dbl</span>(m, ac)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  average    single  complete      ward 
0.9857270 0.9601860 0.9938078 0.9989668 </code></pre>
</div>
</div>
<section id="divisive-hierarchical-clustering" class="level3" style="color: lightyellow">
<h3 class="anchored" data-anchor-id="divisive-hierarchical-clustering">Divisive hierarchical clustering</h3>
</section>
<p>To perform divisive hierarchical clustering with <code>diana()</code>, we simply feed the data frame into the function and specify the dissimilarity measure to be used (i.e.&nbsp;<code>"euclidean"</code>).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="co"># compute divisive hierarchical clustering</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>hc4 <span class="ot">&lt;-</span> <span class="fu">diana</span>(dailyscale[,<span class="fu">c</span>(<span class="st">"lteq"</span>,<span class="st">"posaff"</span>)])</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Divise coefficient; amount of clustering structure found</span></span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>hc4<span class="sc">$</span>dc</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.9929078</code></pre>
</div>
</div>
<section id="determining-the-number-of-clusters" class="level3" style="color: lightyellow">
<h3 class="anchored" data-anchor-id="determining-the-number-of-clusters">Determining the number of clusters</h3>
</section>
<p>Determining the optimal number of clusters is a key step in cluster analysis. Several methods can be used to evaluate the appropriate number of clusters in a dataset. Here are three commonly used methods:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(factoextra) <span class="co"># for cluster visualization</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot cluster results</span></span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">fviz_nbclust</span>(dailyscale[,<span class="fu">c</span>(<span class="st">"lteq"</span>,<span class="st">"posaff"</span>)], <span class="at">FUN =</span> hcut, <span class="at">method =</span> <span class="st">"wss"</span>, </span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a>                   <span class="at">k.max =</span> <span class="dv">10</span>) <span class="sc">+</span></span>
<span id="cb26-6"><a href="#cb26-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"(A) Elbow method"</span>)</span>
<span id="cb26-7"><a href="#cb26-7" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">fviz_nbclust</span>(dailyscale[,<span class="fu">c</span>(<span class="st">"lteq"</span>,<span class="st">"posaff"</span>)], <span class="at">FUN =</span> hcut, <span class="at">method =</span> <span class="st">"silhouette"</span>, </span>
<span id="cb26-8"><a href="#cb26-8" aria-hidden="true" tabindex="-1"></a>                   <span class="at">k.max =</span> <span class="dv">10</span>) <span class="sc">+</span></span>
<span id="cb26-9"><a href="#cb26-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"(B) Silhouette method"</span>)</span>
<span id="cb26-10"><a href="#cb26-10" aria-hidden="true" tabindex="-1"></a>p3 <span class="ot">&lt;-</span> <span class="fu">fviz_nbclust</span>(dailyscale[,<span class="fu">c</span>(<span class="st">"lteq"</span>,<span class="st">"posaff"</span>)], <span class="at">FUN =</span> hcut, <span class="at">method =</span> <span class="st">"gap_stat"</span>, </span>
<span id="cb26-11"><a href="#cb26-11" aria-hidden="true" tabindex="-1"></a>                   <span class="at">k.max =</span> <span class="dv">10</span>) <span class="sc">+</span></span>
<span id="cb26-12"><a href="#cb26-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"(C) Gap statistic"</span>)</span>
<span id="cb26-13"><a href="#cb26-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-14"><a href="#cb26-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Display plots side by side</span></span>
<span id="cb26-15"><a href="#cb26-15" aria-hidden="true" tabindex="-1"></a>gridExtra<span class="sc">::</span><span class="fu">grid.arrange</span>(p1, p2, p3, <span class="at">nrow =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="week6_2_files/figure-html/unnamed-chunk-11-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<section id="dendograms-1" class="level3" style="color: lightyellow">
<h3 class="anchored" data-anchor-id="dendograms-1">Dendograms</h3>
</section>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Construct dendorgram for the Ames housing example</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>hc5 <span class="ot">&lt;-</span> <span class="fu">hclust</span>(d, <span class="at">method =</span> <span class="st">"ward.D2"</span> )</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>dend_plot <span class="ot">&lt;-</span> <span class="fu">fviz_dend</span>(hc5)</span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>dend_data <span class="ot">&lt;-</span> <span class="fu">attr</span>(dend_plot, <span class="st">"dendrogram"</span>)</span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a>dend_cuts <span class="ot">&lt;-</span> <span class="fu">cut</span>(dend_data, <span class="at">h =</span> <span class="dv">8</span>)</span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a><span class="fu">fviz_dend</span>(dend_cuts<span class="sc">$</span>lower[[<span class="dv">2</span>]])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="week6_2_files/figure-html/unnamed-chunk-12-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>In order to identify sub-groups (i.e., clusters), we can cut the dendrogram with <code>cutree()</code>. The height of the cut to the dendrogram controls the number of clusters obtained. Here, we cut our agglomerative hierarchical clustering model into eight clusters. We can see that the concentration of observations are in clusters 1–3.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Ward's method</span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>hc5 <span class="ot">&lt;-</span> <span class="fu">hclust</span>(d, <span class="at">method =</span> <span class="st">"ward.D2"</span> )</span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Cut tree into 4 groups</span></span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a>sub_grp <span class="ot">&lt;-</span> <span class="fu">cutree</span>(hc5, <span class="at">k =</span> <span class="dv">3</span>)</span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-7"><a href="#cb28-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Number of members in each cluster</span></span>
<span id="cb28-8"><a href="#cb28-8" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(sub_grp)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>sub_grp
  1   2   3 
424 693 259 </code></pre>
</div>
</div>
<p>Then we can visualize the clusters formed.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot full dendogram</span></span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a><span class="fu">fviz_dend</span>(</span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>  hc5,</span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">k =</span> <span class="dv">3</span>,</span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">horiz =</span> <span class="cn">TRUE</span>,</span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">rect =</span> <span class="cn">TRUE</span>,</span>
<span id="cb30-7"><a href="#cb30-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">rect_fill =</span> <span class="cn">TRUE</span>,</span>
<span id="cb30-8"><a href="#cb30-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">rect_border =</span> <span class="st">"jco"</span>,</span>
<span id="cb30-9"><a href="#cb30-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">k_colors =</span> <span class="st">"jco"</span>,</span>
<span id="cb30-10"><a href="#cb30-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">cex =</span> <span class="fl">0.1</span></span>
<span id="cb30-11"><a href="#cb30-11" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="week6_2_files/figure-html/unnamed-chunk-14-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>We can also look at the clusters in the bivariate space.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="fu">fviz_cluster</span>(<span class="fu">list</span>(<span class="at">data =</span> dailyscale[,<span class="fu">c</span>(<span class="st">"lteq"</span>,<span class="st">"posaff"</span>)], <span class="at">cluster =</span> sub_grp),</span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>             <span class="at">geom =</span> <span class="st">"point"</span>,</span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a>             <span class="at">ellipse.type =</span> <span class="st">"convex"</span>,</span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a>             <span class="at">palette =</span> <span class="st">"jco"</span>,</span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a>             <span class="at">ggtheme =</span> <span class="fu">theme_minimal</span>())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="week6_2_files/figure-html/unnamed-chunk-15-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<section id="final-thoughts" class="level3">
<h3 class="anchored" data-anchor-id="final-thoughts">Final Thoughts</h3>
<p>Please remember the usual caution. Our intention here has been simple exposure. When using these methods for a paper or project, do the research necessary to engage the method precisely and with good form.</p>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->




</body></html>